File: bindings/java/src/main/java/com/aster/interceptors/AuditInterceptor.java
package com.aster.interceptors;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Logger;

/**
 * Captures structured audit events for requests, responses, and errors.
 *
 * <p>Events are appended to an in-memory sink (a {@code List<Map<String,Object>>}) and logged via
 * {@link java.util.logging.Logger} at FINE level.
 */
public final class AuditInterceptor implements Interceptor {

  private static final Logger DEFAULT_LOGGER = Logger.getLogger(AuditInterceptor.class.getName());

  private final List<Map<String, Object>> sink;
  private final Logger logger;

  /** Creates an audit interceptor with a new empty sink and the default logger. */
  public AuditInterceptor() {
    this(new ArrayList<>(), DEFAULT_LOGGER);
  }

  /**
   * Creates an audit interceptor with the given sink and logger.
   *
   * @param sink mutable list to which audit entries are appended
   * @param logger logger for debug output
   */
  public AuditInterceptor(List<Map<String, Object>> sink, Logger logger) {
    this.sink = sink;
    this.logger = logger != null ? logger : DEFAULT_LOGGER;
  }

  /** Returns the audit event sink. */
  public List<Map<String, Object>> sink() {
    return sink;
  }

  @Override
  public Object onRequest(CallContext ctx, Object request) {
    record("request", ctx, Map.of());
    return request;
  }

  @Override
  public Object onResponse(CallContext ctx, Object response) {
    record("response", ctx, Map.of());
    return response;
  }

  @Override
  public RpcError onError(CallContext ctx, RpcError error) {
    record("error", ctx, Map.of("code", error.code().name(), "message", error.rpcMessage()));
    return error;
  }

  private void record(String event, CallContext ctx, Map<String, Object> extra) {
    Map<String, Object> entry = new HashMap<>();
    entry.put("event", event);
    entry.put("service", ctx.service());
    entry.put("method", ctx.method());
    entry.put("call_id", ctx.callId());
    entry.put("attempt", ctx.attempt());
    entry.put("ts", System.currentTimeMillis() / 1000.0);
    entry.putAll(extra);

    sink.add(entry);
    logger.fine("audit=" + entry);
  }
}

File: bindings/dotnet/src/Aster/Interceptors/AuditInterceptor.cs
namespace Aster.Interceptors;

/// <summary>
/// Captures structured audit events for requests, responses, and errors.
/// Events are appended to a sink (list of dictionaries) and optionally logged.
/// </summary>
public class AuditInterceptor : IInterceptor
{
    /// <summary>The audit event sink. Inspect after calls to review the trail.</summary>
    public List<Dictionary<string, object>> Sink { get; }

    /// <summary>Optional action for logging audit entries (e.g. Console.WriteLine).</summary>
    public Action<Dictionary<string, object>>? Logger { get; set; }

    /// <param name="sink">External sink list, or null to create a new one.</param>
    /// <param name="logger">Optional logging callback.</param>
    public AuditInterceptor(
        List<Dictionary<string, object>>? sink = null,
        Action<Dictionary<string, object>>? logger = null)
    {
        Sink = sink ?? new List<Dictionary<string, object>>();
        Logger = logger;
    }

    public object OnRequest(CallContext ctx, object request)
    {
        Record("request", ctx);
        return request;
    }

    public object OnResponse(CallContext ctx, object response)
    {
        Record("response", ctx);
        return response;
    }

    public RpcError? OnError(CallContext ctx, RpcError error)
    {
        Record("error", ctx,
            ("code", error.Code.ToString()),
            ("message", error.Message));
        return error;
    }

    private void Record(string eventName, CallContext ctx, params (string key, object value)[] extra)
    {
        var entry = new Dictionary<string, object>
        {
            ["event"] = eventName,
            ["service"] = ctx.Service,
            ["method"] = ctx.Method,
            ["call_id"] = ctx.CallId,
            ["attempt"] = ctx.Attempt,
            ["ts"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0,
        };
        foreach (var (key, value) in extra)
        {
            entry[key] = value;
        }
        Sink.Add(entry);
        Logger?.Invoke(entry);
    }
}
